gobblin-modules/gobblin-elasticsearch/src/main/java/org/apache/gobblin/elasticsearch/writer/FutureCallbackHolder.java

```diff
@@ -0,0 +1,126 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.gobblin.elasticsearch.writer;
+
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import org.apache.commons.math3.util.Pair;
+import org.apache.gobblin.writer.GenericWriteResponse;
+import org.apache.gobblin.writer.WriteCallback;
+import org.apache.gobblin.writer.WriteResponse;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.bulk.BulkResponse;
+
+import javax.annotation.Nullable;
+import lombok.Getter;
+
+
+/**
+ * A class to hold Futures and Callbacks to support Async writes
+ */
+public class FutureCallbackHolder {
+
+  @Getter
+  private final ActionListener<BulkResponse> actionListener;
+  private final BlockingQueue<Pair<WriteResponse, Throwable>> writeResponseQueue = new ArrayBlockingQueue<>(1);
+  @Getter
+  private final Future<WriteResponse> future;
+
+
+  public FutureCallbackHolder(final @Nullable WriteCallback callback) {
+    this(callback, null);
+  }
+
+  public FutureCallbackHolder(final @Nullable WriteCallback callback, ResponseErrorLogger responseErrorLogger) {
+
+    this.future = new Future<WriteResponse>() {
+      @Override
+      public boolean cancel(boolean mayInterruptIfRunning) {
+        return false;
+      }
+
+      @Override
+      public boolean isCancelled() {
+        return false;
+      }
+
+      @Override
+      public boolean isDone() {
+        return !writeResponseQueue.isEmpty();
+      }
+
+      @Override
+      public WriteResponse get()
+          throws InterruptedException, ExecutionException {
+        Pair<WriteResponse, Throwable> writeResponseThrowablePair = writeResponseQueue.take();
+        return getWriteResponseorThrow(writeResponseThrowablePair);
+      }
+
+      @Override
+      public WriteResponse get(long timeout, TimeUnit unit)
+          throws InterruptedException, ExecutionException, TimeoutException {
+        Pair<WriteResponse, Throwable> writeResponseThrowablePair = writeResponseQueue.poll(timeout, unit);
+        if (writeResponseThrowablePair == null) {
+          throw new TimeoutException("Timeout exceeded while waiting for future to be done");
```

Comment:
In this timeout case, do we need to invoke the callback?

Reply:
This is a client specifying how long they want to wait on the future... they could decide to issue this call again... so IMO the callback should only be fired when the underlying call to the destination succeeds or fails.